### PR TITLE
nixos/kubo: improve API access

### LINKED ISSUE #223289: services.kubo: ipfs CLI only works under root
### Describe the bug

When running an ipfs/kubo node using `services.kubo.enable = true`, the `ipfs` CLI is not usable unless executed with root privileges. E.g.:

```console
❯ ipfs id
Error: error loading plugins: open /var/lib/ipfs/config: permission denied
❯ sudo -E ipfs id
{
	"ID": ...
}
```

I don't know if that's a bug or the intended behaviour.
There's a NixOS test for kubo, which tests `ipfs ... id`, but only runs it with root privileges:
https://github.com/NixOS/nixpkgs/blob/151981821c03435ee8047f6c1b9367c1a26788ed/nixos/tests/kubo.nix#L30

### Steps To Reproduce
1. `services.kubo.enable = true`
2. `ipfs id` `->` get a "permission denied" error
3. `sudo -E ipfs id`

### Expected behavior

Both commands print the json with node public info.
If step 2 was supposed to fail after all, I expected to see a different description of the module or a comment in the test.

### Notify maintainers

- @Luflosi (`services.kubo` maintainer)
-  @Ericson2314 (edited the nixos test)

As I said, I don't know if this was the intended behaviour. Just in case, I apologize for the noise 🙃 
### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.19, NixOS, 23.05 (Stoat), 23.05.20230314.7067edc`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.14.1`
 - nixpkgs: `/etc/nixpkgs/channels/nixpkgs`


```

### PR DESCRIPTION
**Pull request by Luflosi**
###### Description of changes
Fixes https://github.com/NixOS/nixpkgs/issues/223289.

This allows the ipfs CLI tools to communicate with the daemon when the user is in the ipfs group. Before, the user had to be root or do some annoying and error-prone manual setup to allow this.
I also locked down the security a bit. Before this PR, all users on the system could access the API of the Kubo daemon with some manual setup but with this PR, only users in the ipfs group can talk to the daemon by default.

I feel like the wording in some commit messages, comments and changelog documentation could be improved. Please suggest improvements.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Luflosi** [comment]
If I requested a review from you, I saw that you made a change to the Kubo module or test in the not too distant past.
If you no longer care about Kubo, I apologise for the ping.

**Comment by Ericson2314** [comment]
This looks good to me.And the issue has bugged me for quite some time. Thank you!

**Review by Ericson2314 — APPROVED** [review]

File: nixos/modules/services/network-filesystems/kubo.nix [hunk 1]
```diff
@@ -154,13 +182,18 @@ in
...
+                Multiaddr or array of multiaddrs describing the address to serve the local HTTP API on.
+                In addition to the multiaddrs listed here, the daemon will also listen on a Unix domain socket.
+                To allow the ipfs CLI tools to communicate with the daemon over that socket,
+                add your user to the correct group, e.g. `users.users.alice.extraGroups = [ config.services.kubo.group ];`
+              '';
             };
 
             Addresses.Gateway = mkOption {
-              type = types.str;
+              type = types.oneOf [ types.str (types.listOf types.str) ];
```
Comment (Ericson2314) [inline nixos/modules/services/network-filesystems/kubo.nix:196]: I would consider deprecating the non-list version so we can remove it later.

Comment (Luflosi) [reply nixos/modules/services/network-filesystems/kubo.nix:196]: I'll leave that to someone else. And besides, the Kubo config file also supports both.

**Pull request merged**